README.md
<!--
 * @Author: atdow
 * @Date: 2022-11-01 21:07:59
 * @LastEditors: null
 * @LastEditTime: 2022-11-05 01:37:33
 * @Description: file description
-->

# vscode-vue-jumper

vue文件跳转到文件定义支持。

![](./images/usage-demo.gif)

## 1. 标签跳转支持

支持大驼峰组件、中划线组件。

```html
<my-component></my-component>
<MyComponent></MyComponent>
```

## 2. import相对路径文件跳转支持

```js
import MyComponent form '../../component/MyComponent'
import MyComponent2 form '../../component/MyComponent2.vue'
```

## 3. import别名路径文件跳转支持

```js
import MyComponent form '@/component/MyComponent'
```

默认配置了 `@:src` ，如果有需要，请到插件配置中设置aliasConfigs：

格式： `别名名称:目标路径`

## 4. mixins引入组件注册跳转支持

```html
<template>
    <div>
        <MyComponent />
    </div>
</template>

<script>
    import mixins from "./mixins";
    export default {
        mixins: [mixins],
        props: {},
        data() {
            return {};
        },
        components: {},
    };
</script>
```

```js
// mixins.js
import MyComponent from "./modules/MyComponent";
export default {
    data() {
        return {}
    },
    components: {
        MyComponent
    }
}
```

## 版本

* 2.0.0 增加mixins引入组件注册跳转支持
* 1.3.0-1.5.0 修复匹配错误
* 1.2.0 增加组件重命名跳转
* 1.1.0 支持多workspaceFolders工作区跳转
* 1.0.0 支持基础跳转


CHANGELOG.md
# Change Log

All notable changes to the "vue-jumper" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

* Initial release


src/util.ts
/*
 * @Author: atdow
 * @Date: 2022-10-29 19:56:13
 * @LastEditors: null
 * @LastEditTime: 2022-11-05 17:59:05
 * @Description: file description
 */
const fs = require('fs')
const os = require('os')
const path = require('path')
const vscode = require('vscode')
import { IAliasConfigsItem } from './types'

const util = {
  isWindows: function (): boolean {
    // console.log("fs:", fs);
    const isWindows: boolean = /\\/.test(vscode.workspace.rootPath)
    return isWindows
  },
  getCurrentDir: function (currentFile: string): string {
    let workspaceFolders: string[] = vscode.workspace.workspaceFolders.map((item) => item.uri.path)
    let formatCurrentFile: string = currentFile
    if (this.isWindows()) {
      formatCurrentFile = '/' + currentFile.replace(/\\/g, '/')
    }
    // 当前工作区
    let currentWorkspaceFolder: string = ''
    workspaceFolders.forEach((workspaceFoldersItem) => {
      // 这里的判断应该是无用的，可能是虚拟机导致的问题
      if (formatCurrentFile.startsWith('/w:') || formatCurrentFile.startsWith('/W:')) {
        if (formatCurrentFile.slice(3).indexOf(workspaceFoldersItem.slice(3)) === 0) {
          currentWorkspaceFolder = workspaceFoldersItem
        }
      } else {
        if (formatCurrentFile.indexOf(workspaceFoldersItem) === 0) {
          currentWorkspaceFolder = workspaceFoldersItem
        }
      }
    })
    // 这里多了一层目录，也就是当前文件名
    const currentFilePath: string = formatCurrentFile.slice(currentWorkspaceFolder.length + 1)
    const currentFilePathArr: string[] = currentFilePath.split('/')
    return currentFilePathArr.slice(0, currentFilePathArr.length - 1).join('/') // 去除当前文件名并返回
  },
  /**
   * 从文档文本中查找组件引入代码行
   * @param documentText 文档文本
   * @param componentName 组件名
   * @returns 引入文本行: import MyComponent from "../../MyComponent"
   */
  documentTextFindComponentImportLine(documentText: string, componentName: string): string {
    let importLine: string = ''
    if (documentText.match(/import.+['"]/)) {
      const importArr: string[] = documentText.match(/[^//]import.+['"]/g)
      importLine = importArr.find((item) => item.indexOf(componentName) !== -1) || ''
    }
    return importLine
  },
  documentFindAllImport(
    // document,
    documentText: string = '',
    aliasConfigs: IAliasConfigsItem[] = [],
    currentFilePath: string = ''
  ): object {
    const obj: object = {}
    // console.log("documentText:", documentText);
    if (documentText.match(/import.+['"]/)) {
      const importArr: string[] = documentText.match(/[^//]import.+['"]/g)
      importArr.forEach((importLineItem) => {
        let componentName: string = ''
        const path: string = this.importLineFindOriginImportPath(importLineItem)
        const tagSliceArr: string[] = importLineItem.match(/import.*?from/) || [] // ['import componentName from']
        if (tagSliceArr.length > 0) {
          // ['import', 'componentName', 'from']
          componentName = (tagSliceArr[0].match(/[a-zA-Z0-9_]+/g) || []).find(
            (item) => item.trim() !== 'import' && item.trim() !== 'from'
          )
        }
        if (componentName) {
          obj[componentName] = { originPath: path }
        }
        // console.log("componentName:", componentName);
      })
    }
    this.documentImportObjAddPath(obj, aliasConfigs, currentFilePath)
    // console.log("obj:", obj);
    return obj
  },
  documentImportObjAddPath(
    documentImportObj: object = {},
    aliasConfigs: IAliasConfigsItem[] = [],
    currentFilePath: string = ''
  ) {
    Object.keys(documentImportObj).forEach((key) => {
      let absolutePath: string = ''
      const originImportPath: string = documentImportObj[key].originPath
      const pureOriginImportPath: string = originImportPath.replace(/^[\.]\//, '') // 清除./
      const currentDirPath: string = this.getCurrentDir(currentFilePath)
      // console.log('currentDirPath:', currentDirPath)
      // 带../相对路径
      if (pureOriginImportPath.match(/\.\.\//)) {
        const appendPath: string = pureOriginImportPath.replace(/\.\.\//, '') // 去除../
        const pathMeshArr: string[] = pureOriginImportPath.match(/\.\.\//)
        const currentDirPathArr: string[] = currentDirPath.split('/')
        // 相对目录路径
        const prefixPath: string = currentDirPathArr.slice(0, currentDirPathArr.length - pathMeshArr.length).join('/')
        absolutePath = `${prefixPath}/${appendPath}`
      } else {
        // 当前目录下
        absolutePath = `${currentDirPath}/${pureOriginImportPath}`
      }
      // alias别名替换
      aliasConfigs.forEach((aliasConfigsItem) => {
        if (originImportPath.startsWith(aliasConfigsItem.alias)) {
          absolutePath = originImportPath.replace(new RegExp(`${aliasConfigsItem.alias}`), aliasConfigsItem.target)
        }
      })
      documentImportObj[key].path = absolutePath
    })
  },
  documentFindRegisterComponentsObj(documentText: string = ''): object {
    // const documentText = document.getText();
    /**
      匹配到以下结果:
      components: { 's-component': SComponent, MyComponent }

      components: { 
        // 我是注释
        's-component': SComponent, // 我是注释
        MyComponent 
      }
     */
    const componentsCombine: string[] = documentText.match(/components[\s]?:[\s\S]*?{[\s\S]*?}/g) || []
    // console.log("componentsCombine:", componentsCombine);
    if (componentsCombine && componentsCombine.length > 0) {
      /**
        匹配到以下结果:
        { 's-component': SComponent, MyComponent }

        { 
          // 我是注释
          's-component': SComponent, // 我是注释
          MyComponent 
        }
       */
      const componentObjStrArr: string[] = componentsCombine[0].match(/{[\s\S]*?}/) || []
      let componentObjStr: string = ''
      if (componentObjStrArr && componentObjStrArr.length > 0) {
        componentObjStr = componentObjStrArr[0]
      } else {
        return {}
      }
      // ["s-component: SComponent", "MyComponent"]
      const componentArr: string[] = componentObjStr
        .replace(/(\n+)|(\/{2,}.*?(\r|\n))|(\/\*(\n|.)*?\*\/)/g, '') // 去掉注释
        .replace(/[{}'"]/g, '') // 去掉花括号、'、"
        .split(',')
        .map((item) => item.trim())
        .filter((item) => !!item)
      // console.log("componentArr:", componentArr);
      const registerComponentsObj: object = {}
      componentArr.forEach((item) => {
        const splitArr: string[] = item.split(':')
        if (splitArr.length === 2) {
          registerComponentsObj[splitArr[0].trim()] = splitArr[1].trim()
        } else {
          registerComponentsObj[splitArr[0].trim()] = splitArr[0].trim()
        }
      })
      return registerComponentsObj
      // console.log("componentsObj:", componentsObj);
    }
    return {}
  },
  documentFindMixins(documentText: string = ''): string[] {
    documentText = documentText.replace(/(\n+)|(\/{2,}.*?(\r|\n))|(\/\*(\n|.)*?\*\/)/g, '') // 去掉注释
    // const documentText = document.getText();
    const mixinsCombine: string[] = documentText.match(/mixins[\s]?:[\s\S]*?\[[\s\S]*?\]/g) || []
    // console.log("mixinsCombine:", mixinsCombine);
    if (mixinsCombine && mixinsCombine.length > 0) {
      const mixinsStrArr: string[] = mixinsCombine[0].match(/\[[\s\S]*?\]/) || []
      let mixinsStr: string = ''
      if (mixinsStrArr.length > 0) {
        mixinsStr = mixinsStrArr[0]
        const mixins: string[] = mixinsStr
          .replace(/[\[\]]/g, '') // 去掉[、]
          .split(',')
          .map((item) => item.trim())
          .filter((item) => !!item)
        return mixins
      } else {
        return []
      }
    }
    return []
  },
  /**
   * 将tagName转成大驼峰
   * 主要转换: my-components --> MyComponents
   * @param tagName
   */
  upperCamelCaseTagName: function (tagName: string): string {
    let formatTagName: string = tagName
    if (formatTagName.indexOf('-') !== -1) {
      let myText: string = ''
      const textCharArr: string[] = tagName.split('')
      for (let i = 0; i < textCharArr.length; i++) {
        if (i === 0) {
          myText += textCharArr[i].toUpperCase()
        } else if (textCharArr[i] === '-') {
          textCharArr[i + 1] = textCharArr[i + 1].toUpperCase()
        } else {
          myText += textCharArr[i]
        }
      }
      formatTagName = myText
    }
    return formatTagName
  },
  /**
   * 引入文本行查找原始的引入文件地址
   * @param importLine 引入文本: import MyComponent from "../../MyComponent"
   * @returns ../../MyComponent
   */
  importLineFindOriginImportPath: function (importLine: string): string {
    let originImportPath: string = ''
    // import MyComponent from "../../MyComponent"
    if (importLine.match(/\"([^\"]*)\"/)) {
      const pathArr: string[] = importLine.match(/\"([^\"]*)\"/) || []
      if (pathArr && pathArr.length > 0) {
        originImportPath = pathArr.find((item) => !item.match(/\"/))
      }
    }
    // import MyComponent from '../../MyComponent'
    if (importLine.match(/'([^\']*)'/)) {
      const pathArr: string[] = importLine.match(/'([^\']*)'/) || []
      if (pathArr && pathArr.length > 0) {
        originImportPath = pathArr.find((item) => !item.match(/'/))
      }
    }
    return originImportPath
  },
  /**
   * 弹出错误信息
   */
  showError: function (info: string) {
    vscode.window.showErrorMessage(info)
  },
  /**
   * 弹出提示信息
   */
  showInfo: function (info: string) {
    vscode.window.showInformationMessage(info)
  }
}

module.exports = util


src/extension.ts
/*
 * @Author: atdow
 * @Date: 2017-08-21 14:59:59
 * @LastEditors: null
 * @LastEditTime: 2022-11-05 18:04:44
 * @Description: file description
 */
'use strict'
import * as vscode from 'vscode'
import JumperFileDefinitionProvider from './JumperFileDefinitionProvider'

const languageConfiguration: vscode.LanguageConfiguration = {
  wordPattern: /(\w+((-\w+)+)?)/
}

export function activate(context: vscode.ExtensionContext) {
  const configParams = vscode.workspace.getConfiguration('vue-jumper')
  const supportedLanguages = configParams.get('supportedLanguages') as Array<string>
  const targetFileExtensions = configParams.get('targetFileExtensions') as Array<string>
  const aliasConfigs = configParams.get('aliasConfigs') as Array<string>

  context.subscriptions.push(
    vscode.languages.registerDefinitionProvider(
      supportedLanguages,
      new JumperFileDefinitionProvider(targetFileExtensions, aliasConfigs)
    )
  )

  /* Provides way to get selected text even if there is dash
   * ( must have fot retrieving component name )
   */
  context.subscriptions.push(vscode.languages.setLanguageConfiguration('vue', languageConfiguration))
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/PeekFileDefinitionProvider.ts
/*
 * @Author: atdow
 * @Date: 2017-08-21 14:59:59
 * @LastEditors: null
 * @LastEditTime: 2022-10-29 17:24:18
 * @Description: file description
 */
import * as vscode from "vscode";

export default class PeekFileDefinitionProvider
  implements vscode.DefinitionProvider
{
  targetFileExtensions: string[] = [];

  constructor(targetFileExtensions: string[] = []) {
    this.targetFileExtensions = targetFileExtensions;
  }

  judeLineType(line: String, text: string, document) {
    const lineInfo: any = {
      // type: "import",
      // path: "",
      // simplePath: ""
    };
    if (!line) {
      return lineInfo;
    }
    const pureLine = line.trim();
    // import 类型
    if (pureLine.startsWith("import")) {
      lineInfo.type = "import";
      const { path, simplePath } = importTypeAnalysis(pureLine);
      lineInfo.path = path;
      lineInfo.simplePath = simplePath;
    }
    function importTypeAnalysis(pureLine) {
      let path = "";
      let simplePath = "";
      // "xxx"
      if (pureLine.match(/\"([^\"]*)\"/)) {
        const pathArr = pureLine.match(/\"([^\"]*)\"/);
        if (pathArr && pathArr.length > 0) {
          path = pathArr.find((item) => !item.match(/\"/));
        }
      }
      // 'xxx'
      if (pureLine.match(/'([^\']*)'/)) {
        const pathArr = pureLine.match(/'([^\']*)'/);
        if (pathArr && pathArr.length > 0) {
          path = pathArr.find((item) => !item.match(/'/));
        }
      }
      simplePath = path.replace(/\.\.\//, "").replace(/\.\//, "");
      return { path, simplePath };
    }
    // 标签类型
    if (pureLine.startsWith("<")) {
      lineInfo.type = "tag";
      let formatText = text;
      // my-components(<my-components></my-components>形式标签)转MyComponents(<MyComponents></MyComponents>形式标签)
      if (formatText.indexOf("-") !== -1) {
        let myText = "";
        const textCharArr = text.split("");
        for (let i = 0; i < textCharArr.length; i++) {
          if (i === 0) {
            myText += textCharArr[i].toUpperCase();
          } else if (textCharArr[i] === "-") {
            textCharArr[i + 1] = textCharArr[i + 1].toUpperCase();
          } else {
            myText += textCharArr[i];
          }
        }
        formatText = myText;
      }
      // console.log("tag-----------------:");
      // console.log("formatText:", formatText);
      const json = document.getText();
      // 查找标签引入地方
      if (json.match(/import.+'/)) {
        const importArr = json.match(/import.+'/g);
        const importLine = importArr.find(
          (item) => item.indexOf(formatText) !== -1
        );
        // console.log("importLine:", importLine);
        if (importLine) {
          const { path, simplePath } = importTypeAnalysis(importLine.trim());
          lineInfo.path = path;
          lineInfo.simplePath = simplePath;
        }
      }
    }
    return lineInfo;
  }

  getComponentName(position: vscode.Position, document): String[] {
    const doc = vscode.window.activeTextEditor.document;
    const selection = doc.getWordRangeAtPosition(position);
    const selectedText = doc.getText(selection);
    let lineText = doc.lineAt(position).text;
    const lineInfo = this.judeLineType(lineText, selectedText, document);
    // console.log("lineText:", lineText);
    console.log("lineInfo:", lineInfo);
    let possibleFileNames = [],
      altName = "";
    if (lineInfo.type === "import") {
      const simplePath = lineInfo.simplePath;
      possibleFileNames.push(simplePath + ".vue");
      possibleFileNames.push(simplePath + "/index.vue");
    } else if (lineInfo.type === "tag") {
      const simplePath = lineInfo.simplePath;
      possibleFileNames.push(simplePath + ".vue");
      possibleFileNames.push(simplePath + "/index.vue");
    }

    // console.log('position:',position)
    // console.log('selection:',selection)
    // console.log('selectedText:',selectedText)
    // console.log('rowText:',rowText)
    // selectedText.match(/\w+/g).forEach(str => {
    //   return altName += str[0].toUpperCase() + str.substring(1);
    // })
    // console.log("this.targetFileExtensions:", this.targetFileExtensions);

    this.targetFileExtensions.forEach((ext) => {
      possibleFileNames.push(selectedText + ext);
      possibleFileNames.push(selectedText + "/index" + ext);
      possibleFileNames.push(altName + ext);
      possibleFileNames.push(altName + "/index" + ext);
    });

    return possibleFileNames;
  }

  searchFilePath(fileName: String): Thenable<vscode.Uri[]> {
    return vscode.workspace.findFiles(`**/${fileName}`, "**/node_modules"); // Returns promise
  }

  provideDefinition(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken
  ): Promise<vscode.Location | vscode.Location[]> {
    let filePaths = [];
    const componentNames = this.getComponentName(position, document);
    // console.log("componentNames:", componentNames);
    // console.log("this.searchFilePath:", this.searchFilePath);
    // console.log("filePaths:", filePaths);
    const searchPathActions = componentNames.map(this.searchFilePath);
    const searchPromises = Promise.all(searchPathActions); // pass array of promises

    return searchPromises.then(
      (paths) => {
        filePaths = [].concat.apply([], paths);

        if (filePaths.length) {
          let allPaths = [];
          filePaths.forEach((filePath) => {
            allPaths.push(
              new vscode.Location(
                vscode.Uri.file(`${filePath.path}`),
                new vscode.Position(0, 1)
              )
            );
          });
          return allPaths;
        } else {
          return undefined;
        }
      },
      (reason) => {
        return undefined;
      }
    );
  }
}


src/JumperFileDefinitionProvider.ts
/*
 * @Author: atdow
 * @Date: 2017-08-21 14:59:59
 * @LastEditors: null
 * @LastEditTime: 2022-11-05 19:34:32
 * @Description: file description
 */
import * as vscode from 'vscode'
const util = require('./util')
const fs = require('fs')
import { IAliasConfigsItem, ILineInfo } from './types'

export default class JumperFileDefinitionProvider implements vscode.DefinitionProvider {
  targetFileExtensions: string[] = []
  aliasConfigs: IAliasConfigsItem[] = []

  constructor(targetFileExtensions: string[] = [], aliasConfigs: string[] = []) {
    this.targetFileExtensions = targetFileExtensions
    aliasConfigs.forEach((aliasConfigsItem) => {
      try {
        const aliasConfigsItemArr: string[] = aliasConfigsItem.split(':')
        if (aliasConfigsItemArr && aliasConfigsItemArr.length === 2) {
          this.aliasConfigs.push({
            alias: aliasConfigsItemArr[0],
            target: aliasConfigsItemArr[1]
          })
        }
      } catch (error) {
        // console.log("aliasConfigs:", aliasConfigs);
      }
    })
  }

  async judeLineType(line: String, keyword: string, document: vscode.TextDocument): Promise<ILineInfo> {
    const that = this
    const lineInfo: ILineInfo = {
      type: '',
      path: '',
      originPath: ''
    }
    if (!line) {
      return lineInfo
    }
    const pureLine: string = line.trim()
    const importObj: object = util.documentFindAllImport(
      document.getText(),
      that.aliasConfigs,
      document.uri.fsPath
      // (document.uri ? document.uri : document).fsPath
    )
    // console.log('importObj:', importObj)
    const registerComponentsObj: object = util.documentFindRegisterComponentsObj(document.getText()) || {}
    // import 类型
    if (pureLine.startsWith('import')) {
      lineInfo.type = 'import'
      this.componentNameInImportObjUpdateLineInfo(keyword, importObj, lineInfo)
    }
    // 标签类型
    if (pureLine.startsWith('<')) {
      lineInfo.type = 'tag'
      let searchComponentName: string = util.upperCamelCaseTagName(keyword)
      // 直接从importObj中查找
      this.componentNameInImportObjUpdateLineInfo(searchComponentName, importObj, lineInfo)
      // 从components中查找(组件重命名情况) components: { RenameMyComponent: MyComponent, 's-my-component2': MyComponent2 }
      if (!lineInfo.path) {
        Object.keys(registerComponentsObj).forEach((key) => {
          if (key === searchComponentName || key === keyword) {
            searchComponentName = registerComponentsObj[key]
          }
        })
        this.componentNameInImportObjUpdateLineInfo(searchComponentName, importObj, lineInfo)
      }
      // 从mixins中找
      if (!lineInfo.path) {
        const mixins: string[] = util.documentFindMixins(document.getText()) || []
        // console.log('mixins:', mixins)
        for (let i = mixins.length - 1; i >= 0; i--) {
          // 从后面往前找，如果找到了就不再找了
          if (lineInfo.path) {
            break
          }
          let mixinsPath: string = importObj[mixins[i]].path
          const mixinsPathArr: Thenable<vscode.Uri[]>[] = []
          if (!mixinsPath.endsWith('.js') && !mixinsPath.endsWith('.ts')) {
            mixinsPathArr.push(this.searchFilePath(`${mixinsPath}.js`))
            mixinsPathArr.push(this.searchFilePath(`${mixinsPath}.ts`))
          } else {
            mixinsPathArr.push(this.searchFilePath(`${mixinsPath}`))
          }
          const mixinsFilePathArr: Array<vscode.Uri[]> = (await Promise.all(mixinsPathArr)) || []
          mixinsFilePathArr.forEach((resItem) => {
            if (resItem.length === 0) {
              return
            }
            const mixinsFilePath: vscode.Uri = resItem[0]
            let readFileSyncFormatFilePath: string = mixinsFilePath.path
            if (util.isWindows()) {
              // /c:/code/xxx/src/views/mixins1.js ==> c://code//xxx//src//views//mixins1.js
              readFileSyncFormatFilePath = readFileSyncFormatFilePath.slice(1).replace(/\//g, '//')
            }
            let file: string = fs.readFileSync(readFileSyncFormatFilePath, { encoding: 'utf-8' })
            if (!file) {
              return
            }
            let documentFindAllImportFormatPath: string = mixinsFilePath.path
            if (util.isWindows()) {
              // /c:/code/xxx/src/views/mixins1.js => c:/code/xxx/src/views/mixins1.js 为了让documentFindAllImport方法保持一致
              documentFindAllImportFormatPath = documentFindAllImportFormatPath.slice(1)
            }
            const mixinsFileImportObj: object = util.documentFindAllImport(
              file,
              that.aliasConfigs,
              documentFindAllImportFormatPath
            )
            const mixinsFileRegisterComponentsObj: object = util.documentFindRegisterComponentsObj(file) || {}
            let searchComponentName: string = util.upperCamelCaseTagName(keyword)
            Object.keys(mixinsFileRegisterComponentsObj).forEach((key) => {
              if (key === searchComponentName || key === keyword) {
                searchComponentName = mixinsFileRegisterComponentsObj[key]
              }
            })
            this.componentNameInImportObjUpdateLineInfo(searchComponentName, mixinsFileImportObj, lineInfo)
          })
        }
      }
    }
    return lineInfo
  }
  /**
   * 从importObj找到对应的componentName信息，并用来更新lineInfo
   * @param componentName
   * @param importObj
   * @param lineInfo
   */
  componentNameInImportObjUpdateLineInfo(componentName: string, importObj: object, lineInfo: ILineInfo) {
    Object.keys(importObj).forEach((key) => {
      if (key === componentName) {
        lineInfo.originPath = importObj[componentName].originPath
        lineInfo.path = importObj[componentName].path
      }
    })
  }

  getComponentName(position: vscode.Position, document: vscode.TextDocument): Promise<string[]> {
    const doc: vscode.TextDocument = vscode.window.activeTextEditor.document
    const selection: vscode.Range = doc.getWordRangeAtPosition(position)
    const selectedText: string = doc.getText(selection)
    let lineText: string = doc.lineAt(position).text
    return this.judeLineType(lineText, selectedText, document)
      .then((res: ILineInfo) => {
        const { type, path, originPath } = res
        let possibleFileNames: string[] = []
        if (type === 'import' || type === 'tag') {
          possibleFileNamesAdd(path)
        }
        function possibleFileNamesAdd(originPath) {
          // 通过常规的方法都无法找到，只能退而其次地模糊去全局文件中找
          if (!path) {
            const upperCamelCaseTagName = util.upperCamelCaseTagName(selectedText)
            possibleFileNames.push(upperCamelCaseTagName + '.vue')
            possibleFileNames.push(upperCamelCaseTagName + '/index.vue')
            possibleFileNames.push(upperCamelCaseTagName + '/index.js')
            possibleFileNames.push(selectedText + '.vue')
            possibleFileNames.push(selectedText + '/index.vue')
            possibleFileNames.push(selectedText + '/index.js')
            return
          }
          if (!path.endsWith('.vue')) {
            possibleFileNames.push(path + '.vue')
            possibleFileNames.push(path + '/index.vue')
          }
          if (!path.endsWith('.js')) {
            possibleFileNames.push(path + '.js')
            possibleFileNames.push(path + '/index.js')
          }
          if (!path.endsWith('.jsx')) {
            possibleFileNames.push(path + '.jsx')
            possibleFileNames.push(path + '/index.jsx')
          }
          possibleFileNames.push(path)
        }

        return possibleFileNames
      })
      .catch(() => {
        return []
      })
  }

  searchFilePath(fileName: String): Thenable<vscode.Uri[]> {
    return vscode.workspace.findFiles(`**/${fileName}`, '**/node_modules') // Returns promise
  }

  provideDefinition(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken
  ): Promise<vscode.Location | vscode.Location[]> {
    let filePaths: vscode.Uri[] = []
    return this.getComponentName(position, document).then((componentNames) => {
      // console.log("componentNames:", componentNames);
      const searchPathActions: Thenable<vscode.Uri[]>[] = componentNames.map(this.searchFilePath)
      const searchPromises = Promise.all(searchPathActions) // pass array of promises
      return searchPromises.then(
        (paths) => {
          filePaths = [].concat.apply([], paths)

          if (filePaths.length) {
            let allPaths: vscode.Location[] = []
            filePaths.forEach((filePath) => {
              allPaths.push(new vscode.Location(vscode.Uri.file(`${filePath.path}`), new vscode.Position(0, 1)))
            })
            return allPaths
          } else {
            return undefined
          }
        },
        (reason) => {
          return undefined
        }
      )
    })
  }
}
